[
    {
        "content": "<p>Hi. I am not sure if this problem I am facing is a result of API misuse.</p>\n<p>I am following this guide: <a href=\"https://github.com/rust-lang/rustc-dev-guide/blob/master/src/rustc-driver-getting-diagnostics.md\">https://github.com/rust-lang/rustc-dev-guide/blob/master/src/rustc-driver-getting-diagnostics.md</a></p>\n<p>My program enters the<code>compiler.enter(|queries| {</code> closure, but crashes when I call <code>queries.global_ctxt()</code> (I am not doing <code>.unwrap()</code>). I tried using <code>queries.global_ctxt().map_err(|err|</code> for printing the error, but it's not even reaching that part.</p>\n<p>I don't have a way to easily extract this section of the code at the moment - too many parameters and the setup is complicated in general. I was wondering if you have hints of what could go wrong for that method specifically.</p>\n<div class=\"codehilite\"><pre><span></span><code>rustc 1.51.0-nightly (b12290861 2021-01-29)\n\n&gt; rustup component   list | grep installed\ncargo-x86_64-unknown-linux-gnu (installed)\nclippy-x86_64-unknown-linux-gnu (installed)\nllvm-tools-preview-x86_64-unknown-linux-gnu (installed)\nrust-docs-x86_64-unknown-linux-gnu (installed)\nrust-src (installed)\nrust-std-x86_64-unknown-linux-gnu (installed)\nrustc-x86_64-unknown-linux-gnu (installed)\nrustc-dev-x86_64-unknown-linux-gnu (installed)\nrustfmt-x86_64-unknown-linux-gnu (installed)\n</code></pre></div>",
        "id": 226261942,
        "sender_full_name": "resolritter",
        "timestamp": 1613243664
    },
    {
        "content": "<p>To reinforce the problem above, if I don't call <code>queries.global_ctxt()</code> then the program does not crash. I am still able to do other things inside of the closure.</p>",
        "id": 226262449,
        "sender_full_name": "resolritter",
        "timestamp": 1613244448
    },
    {
        "content": "<p>Do you have a backtrace of the crash?</p>",
        "id": 226262850,
        "sender_full_name": "cjgillot",
        "timestamp": 1613245010
    },
    {
        "content": "<p><span class=\"user-mention silent\" data-user-id=\"248906\">cjgillot</span> <a href=\"#narrow/stream/182449-t-compiler.2Fhelp/topic/queries.2Eglobal_ctxt().20crashes/near/226262850\">said</a>:</p>\n<blockquote>\n<p>Do you have a backtrace of the crash?</p>\n</blockquote>\n<p>Trying to get it right now. This code is running as a sub-process inside of a server so I don't get the stderr/stdout easily.<br>\nI'll post it here if I manage.</p>",
        "id": 226263509,
        "sender_full_name": "resolritter",
        "timestamp": 1613245990
    },
    {
        "content": "<p>Nope. I tried setting <code>RUST_BACKTRACE</code>, <code>RUST_LIB_BACKTRACE</code> and everything else I am aware of but had no luck getting the crash output (if it even is there).</p>",
        "id": 226264742,
        "sender_full_name": "resolritter",
        "timestamp": 1613247563
    },
    {
        "content": "<p>how does it crash? if it doesn't panic/ICE then you'll have to use LLDB/GDB to obtain a backtrace</p>",
        "id": 226265194,
        "sender_full_name": "Jonas Schievink  [he/him]",
        "timestamp": 1613248152
    },
    {
        "content": "<p>Thanks for the <code>lldb</code> tip. I'm not familiar with systems programming tools and its ecosystem, having only ever used Rust, and only a few times for small personal projects.</p>\n<p>It was my first time using <code>lldb</code> so I didn't know what to look for, but here's my guess. I attached to the running process, interrupted it and got to this point</p>\n<div class=\"codehilite\"><pre><span></span><code>(lldb) thread backtrace\n* thread #1, name = &#39;program&#39;, stop reason = signal SIGSTOP\n  * frame #0: 0x00007fc77381d9b7 libpthread.so.0`__pthread_clockjoin_ex + 551\n    frame #1: 0x00007fc773931c3d libstd-8ca495943bae7405.so`std::sys::unix::thread::Thread::join::h561ef9aaa0a7b236 at thread.rs:177:23\n    frame #2: 0x000055a7288a57c9 program`std::thread::JoinInner$LT$T$GT$::join::h8dff95540a161a04 + 73\n    frame #3: 0x000055a7288a5902 program`std::thread::JoinHandle$LT$T$GT$::join::h60973e00ba84c829 + 18\n    frame #4: 0x000055a72858bcaf program`rustc_interface::util::scoped_thread::h8896c3cc6a978e8a + 383\n    frame #5: 0x000055a72858c06c program`rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals::h86846b17f597f0f9 + 508\n    frame #6: 0x000055a72812a36e program`rustc_interface::interface::run_compiler::hf85adb2f5fc2180e + 1374\n</code></pre></div>\n<p>I interrupted and continued a few times, but it seems to be stuck in that place. Then I realized that must be the reason why I wasn't seeing the crashes - the program is still running, but it's stuck in that <code>Thread::join</code>.</p>",
        "id": 226269170,
        "sender_full_name": "resolritter",
        "timestamp": 1613252505
    },
    {
        "content": "<p>Should be this: <a href=\"https://github.com/rust-lang/rust/blob/3c10a880eca60379343a6c6d19dd5bda38ead55d/compiler/rustc_interface/src/util.rs#L128\">https://github.com/rust-lang/rust/blob/3c10a880eca60379343a6c6d19dd5bda38ead55d/compiler/rustc_interface/src/util.rs#L128</a></p>",
        "id": 226269289,
        "sender_full_name": "resolritter",
        "timestamp": 1613252616
    },
    {
        "content": "<div class=\"codehilite\"><pre><span></span><code>$ man 7 signal\nSIGSTOP   17,19,23    Stop    Stop process\nThe signals SIGKILL and SIGSTOP cannot be caught, blocked, or ignored.\n</code></pre></div>\n<p>This seems unlikely to be a bug in the compiler itself.</p>",
        "id": 226269297,
        "sender_full_name": "Joshua Nelson",
        "timestamp": 1613252629
    },
    {
        "content": "<p><span class=\"user-mention silent\" data-user-id=\"232545\">Joshua Nelson</span> <a href=\"#narrow/stream/182449-t-compiler.2Fhelp/topic/queries.2Eglobal_ctxt().20crashes/near/226269297\">said</a>:</p>\n<blockquote>\n<div class=\"codehilite\"><pre><span></span><code>$ man 7 signal\nSIGSTOP   17,19,23    Stop    Stop process\nThe signals SIGKILL and SIGSTOP cannot be caught, blocked, or ignored.\n</code></pre></div>\n<p>This seems unlikely to be a bug in the compiler itself.</p>\n</blockquote>\n<p>that's because I interrupted the process manually on LLDB, right?</p>",
        "id": 226269342,
        "sender_full_name": "resolritter",
        "timestamp": 1613252659
    },
    {
        "content": "<p>what do you mean by \"crashes when I call queries.global_ctxt()\"? What output do you see?</p>",
        "id": 226269352,
        "sender_full_name": "Joshua Nelson",
        "timestamp": 1613252683
    },
    {
        "content": "<p><span class=\"user-mention silent\" data-user-id=\"232545\">Joshua Nelson</span> <a href=\"#narrow/stream/182449-t-compiler.2Fhelp/topic/queries.2Eglobal_ctxt().20crashes/near/226269352\">said</a>:</p>\n<blockquote>\n<p>what do you mean by \"crashes when I call queries.global_ctxt()\"? What output do you see?</p>\n</blockquote>\n<p>if the findings about <code>Thread::join</code> hanging are on the right track, then it's not crashing; it's instead getting stuck in that place</p>",
        "id": 226269446,
        "sender_full_name": "resolritter",
        "timestamp": 1613252797
    },
    {
        "content": "<p>I think the backtrace is for the wrong thread. I don't know how to get it for another in lldb, but in gdb it's <code>thread apply all where</code>.</p>",
        "id": 226269606,
        "sender_full_name": "Joshua Nelson",
        "timestamp": 1613253019
    },
    {
        "content": "<p><code>gdb</code> tip was very helpful. I'll spend more time investigating until I have a better idea of the problem. Thanks.</p>",
        "id": 226272024,
        "sender_full_name": "resolritter",
        "timestamp": 1613254579
    },
    {
        "content": "<p>here's a gdb session</p>\n<div class=\"codehilite\"><pre><span></span><code>(gdb) break *0x0000558dcb2c14b0\nBreakpoint 2 at 0x558dcb2c14b0\n\n(gdb) info break\nNum     Type           Disp Enb Address            What\n2       breakpoint     keep y   0x0000558dcb2c14b0 &lt;rustc_interface::queries::&lt;impl rustc_interface::interface::Compiler&gt;::enter&gt;\n\n(gdb) continue\nContinuing.\n[New Thread 0x7fadcdde4640 (LWP 2357962)]\n[Switching to Thread 0x7fadcdde4640 (LWP 2357962)]\nThread 18 &quot;rustc&quot; hit Breakpoint 2, 0x0000558dcb2c14b0 in rustc_interface::queries::&lt;impl rustc_interface::interface::Compiler&gt;::enter ()\n\n(gdb) frame\n#0  0x0000558dcb2c1a38 in rustc_interface::queries::&lt;impl rustc_interface::interface::Compiler&gt;::enter ()\n\n(gdb) step\nSingle stepping until exit from function _ZN15rustc_interface7queries54_$LT$impl$u20$rustc_interface..interface..Compiler$GT$5enter17h002bd85f7bf03d47E,\nwhich has no line number information.\n0x00007fadd4dadfcc in _Unwind_Resume (exc=0x7fad74097370) at /build/gcc/src/gcc/libgcc/unwind.inc:235\n235 /build/gcc/src/gcc/libgcc/unwind.inc: No such file or directory.\n\n(gdb) backtrace\n#0  0x00007fadd4dadfcc in _Unwind_Resume (exc=0x7fad74097370) at /build/gcc/src/gcc/libgcc/unwind.inc:235\n#1  0x0000558dcb2c1986 in rustc_interface::queries::&lt;impl rustc_interface::interface::Compiler&gt;::enter ()\n#2  0x0000558dcb459aaa in program::main_loop::&lt;impl program::global_state::GlobalState&gt;::on_notification::{{closure}}::{{closure}}::{{closure}} ()\n#3  0x0000558dcb2c29a0 in rustc_interface::interface::create_compiler_and_run::{{closure}} ()\n#4  0x0000558dcb2c0a95 in rustc_span::with_source_map ()\n#5  0x0000558dcb2c2638 in rustc_interface::interface::create_compiler_and_run ()\n#6  0x0000558dcb2c2114 in rustc_interface::interface::run_compiler::{{closure}} ()\n#7  0x0000558dcb526f5f in rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals::{{closure}}::{{closure}} ()\n#8  0x0000558dcb77e268 in scoped_tls::ScopedKey&lt;T&gt;::set ()\n#9  0x0000558dcb2c0c8b in rustc_span::with_session_globals ()\n#10 0x0000558dcb526e9b in rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals::{{closure}} ()\n#11 0x0000558dcb526b8b in rustc_interface::util::scoped_thread::{{closure}} ()\n#12 0x0000558dcb7d55ea in std::sys_common::backtrace::__rust_begin_short_backtrace ()\n#13 0x0000558dcb5292aa in std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} ()\n#14 0x0000558dcb83397a in &lt;std::panic::AssertUnwindSafe&lt;F&gt; as core::ops::function::FnOnce&lt;()&gt;&gt;::call_once ()\n#15 0x0000558dcb835844 in std::panicking::try::do_call ()\n#16 0x0000558dcb8364fd in __rust_try ()\n#17 0x0000558dcb83458e in std::panicking::try ()\n#18 0x0000558dcb833daa in std::panic::catch_unwind ()\n</code></pre></div>\n<p>I'm confused by the backtrace. Is it already panicking at this point or trying to \"catch\" a panic that could happen?</p>",
        "id": 226278176,
        "sender_full_name": "resolritter",
        "timestamp": 1613264312
    },
    {
        "content": "<p>note: <code>rustc_interface::queries::&lt;impl rustc_interface::interface::Compiler&gt;::enter</code> is <code>compiler.enter(|queries| {</code> from <a href=\"https://github.com/rust-lang/rustc-dev-guide/blob/master/src/rustc-driver-getting-diagnostics.md\">https://github.com/rust-lang/rustc-dev-guide/blob/master/src/rustc-driver-getting-diagnostics.md</a></p>",
        "id": 226278251,
        "sender_full_name": "resolritter",
        "timestamp": 1613264411
    },
    {
        "content": "<p>I am stil confused why you think global_ctxt() crashes</p>",
        "id": 226278588,
        "sender_full_name": "Joshua Nelson",
        "timestamp": 1613264884
    },
    {
        "content": "<p>it would be helpful if you could post the output from a normal run without gdb</p>",
        "id": 226278600,
        "sender_full_name": "Joshua Nelson",
        "timestamp": 1613264906
    },
    {
        "content": "<p><span class=\"user-mention silent\" data-user-id=\"232545\">Joshua Nelson</span> <a href=\"#narrow/stream/182449-t-compiler.2Fhelp/topic/queries.2Eglobal_ctxt().20crashes/near/226278600\">said</a>:</p>\n<blockquote>\n<p>it would be helpful if you could post the output from a normal run without gdb</p>\n</blockquote>\n<p>I'm unable to do that because this program is running as a sub-process inside of a server. This closure specifically is inside of a separate thread from the main one and there's no communication between the two. It's like a \"task\" thread that executes something and finishes.</p>",
        "id": 226279019,
        "sender_full_name": "resolritter",
        "timestamp": 1613265631
    },
    {
        "content": "<p><span class=\"user-mention silent\" data-user-id=\"232545\">Joshua Nelson</span> <a href=\"#narrow/stream/182449-t-compiler.2Fhelp/topic/queries.2Eglobal_ctxt().20crashes/near/226278588\">said</a>:</p>\n<blockquote>\n<p>I am stil confused why you think global_ctxt() crashes</p>\n</blockquote>\n<p>You're right that I don't know if it crashes. Although I'm sure something goes wrong when I call it.</p>",
        "id": 226279036,
        "sender_full_name": "resolritter",
        "timestamp": 1613265672
    },
    {
        "content": "<p>A GDB session going all the way to the point where it starts popping the stacks</p>\n<div class=\"codehilite\"><pre><span></span><code>(gdb) bt\n#0  0x00007f1b53014c2e in rustc_parse::new_parser_from_file ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n#1  0x00007f1b53014532 in rustc_parse::parse_crate_from_file ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n#2  0x00007f1b527e6037 in rustc_interface::passes::parse ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n#3  0x00007f1b527fc486 in rustc_interface::queries::Queries::parse ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n#4  0x00007f1b527fcc0e in rustc_interface::queries::Queries::crate_name ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n#5  0x00007f1b527fe1b3 in rustc_interface::queries::Queries::global_ctxt ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n\n(gdb) step\nSingle stepping until exit from function _ZN11rustc_parse20new_parser_from_file17h8c1a920d3876777bE,\nwhich has no line number information.\n[New Thread 0x7f1b48cd2640 (LWP 2436154)]\n[Thread 0x7f1b4a4e7640 (LWP 2411348) exited]\n0x00007f1b53014532 in rustc_parse::parse_crate_from_file ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n\n(gdb) step\nSingle stepping until exit from function _ZN11rustc_parse21parse_crate_from_file17h3f231bc28cec6087E,\nwhich has no line number information.\n0x00007f1b527e6037 in rustc_interface::passes::parse ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n\n(gdb) step\nSingle stepping until exit from function _ZN15rustc_interface6passes5parse17hc864bcb485ab9aa3E,\nwhich has no line number information.\n0x00007f1b527fc486 in rustc_interface::queries::Queries::parse ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n\n(gdb) step\nSingle stepping until exit from function _ZN15rustc_interface7queries7Queries5parse17h04b6f63b2ef4a9a2E,\nwhich has no line number information.\n0x00007f1b527fcc0e in rustc_interface::queries::Queries::crate_name ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n\n(gdb) step\nSingle stepping until exit from function _ZN15rustc_interface7queries7Queries10crate_name17h249b66c451e7bd9aE,\nwhich has no line number information.\n0x00007f1b527fe1b3 in rustc_interface::queries::Queries::global_ctxt ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n</code></pre></div>\n<p>I'm seeing this <code>_Unwind_Resume</code> pop up when I step out of <code>global_ctxt</code>.</p>\n<div class=\"codehilite\"><pre><span></span><code>(gdb)\nSingle stepping until exit from function _ZN15rustc_interface7queries7Queries11global_ctxt17h30ca19b158936419E,\nwhich has no line number information.\n0x00007f1b4fa9afcc in _Unwind_Resume (exc=0x7f1b080bead0) at /build/gcc/src/gcc/libgcc/unwind.inc:235\n235 /build/gcc/src/gcc/libgcc/unwind.inc: No such file or directory.\n\n(gdb) bt\n#0  0x00007f1b4fa9afcc in _Unwind_Resume (exc=0x7f1b080bead0) at /build/gcc/src/gcc/libgcc/unwind.inc:235\n#1  0x00007f1b527feacb in rustc_interface::queries::Queries::global_ctxt ()\n   from /home/user/.rustup/toolchains/nightly-2021-01-30-x86_64-unknown-linux-gnu/lib/librustc_driver-a7396df6edcbeec7.so\n</code></pre></div>\n<p>Then I get more unwind-named steps after that</p>\n<div class=\"codehilite\"><pre><span></span><code>uw_init_context_1\nuw_frame_state_for\n_Unwind_IsSignalFrame\n_Unwind_Find_registered_FDE\n_Unwind_Find_FDE\nuw_frame_state_for\nuw_init_context_1\n_Unwind_Resume\n\n_Unwind_DebugHook\n_Unwind_RaiseException\n</code></pre></div>",
        "id": 226279501,
        "sender_full_name": "resolritter",
        "timestamp": 1613266468
    },
    {
        "content": "<p>If I just call <code>compiler.enter(|queries| {</code> without <code>global_ctxt</code>, I'm able to do whatever I want inside of closure or outside of it after that. So naively, it has something to do with <code>global_ctxt</code>. I'm not doing any <code>unsafe</code> or <code>unwrap</code> anywhere in my code, so it's unclear what could go wrong from my side.</p>\n<p>Forgive me if I am being annoying, but I don't see another way to reason about this.</p>",
        "id": 226279577,
        "sender_full_name": "resolritter",
        "timestamp": 1613266573
    },
    {
        "content": "<p>I'll try some older nightly releases to see if it behaves the same</p>",
        "id": 226279595,
        "sender_full_name": "resolritter",
        "timestamp": 1613266629
    },
    {
        "content": "<p>I'm sorry but I really don't know how to help if you can't describe what behavior is going wrong</p>",
        "id": 226285522,
        "sender_full_name": "Joshua Nelson",
        "timestamp": 1613277576
    },
    {
        "content": "<p>What do you expect to see that's different from what's actually happening?</p>",
        "id": 226285535,
        "sender_full_name": "Joshua Nelson",
        "timestamp": 1613277599
    },
    {
        "content": "<p>I decided I'd stop working on this for now. Thanks for the help.</p>",
        "id": 226320721,
        "sender_full_name": "resolritter",
        "timestamp": 1613331209
    }
]